src/Cli/func/CancelKeyHandler.cs

```diff
@@ -0,0 +1,97 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using System.Runtime.InteropServices;
+using Azure.Functions.Cli.Interfaces;
+using Colors.Net;
+
+namespace Azure.Functions.Cli;
+
+internal static class CancelKeyHandler
+{
+    private static readonly object _lock = new();
+    private static IProcessManager _processManager;
+    private static IConsoleReader _consoleReader;
+    private static Action _onFirstCancel;
+    private static Action _onSecondCancel;
+    private static int _cancelKeyPressCount = 0;
+
+    internal static void Register(IProcessManager processManager, IConsoleReader consoleReader, Action onFirstCancel = null, Action onSecondCancel = null)
+    {
+        _processManager = processManager ?? throw new ArgumentNullException(nameof(processManager));
+        _consoleReader = consoleReader ?? throw new ArgumentNullException(nameof(consoleReader));
+
+        _onFirstCancel = onFirstCancel ?? (() => { });
+        _onSecondCancel = onSecondCancel ?? (() => { });
+
+        Console.CancelKeyPress += HandleCancelKeyPress;
+    }
+
+    internal static void HandleCancelKeyPress(object sender, ConsoleCancelEventArgs e)
+    {
+        lock (_lock)
+        {
+            _cancelKeyPressCount++;
+
+            if (_cancelKeyPressCount == 1)
+            {
+                var message = RuntimeInformation.IsOSPlatform(OSPlatform.Windows)
+                            ? "Press 'q' to force exit."
```

Comment:
@mattchenderson I'm not sure how I feel about this but it works, and we can't depend on Ctrl+C again (thank you windows).

Reply:
I'm similarly not sure how I feel about this. It feels fine for non-Windows; Ctrl-C again is a understandable gesture in response to it not taking effect the first time. The console message helps.

What is the Windows behavior here that makes it different?

Reply:
Per offline thread with Lilian, the issue there is that the second Ctrl+C wouldn't register, which seemed intentional.

But the more I think about it, the more I'm unsure about introducing the extra key command here. There's something to the repeat being an extra signal of intent on its own. I know I said the console message helps in my last comment, but I can't say I love it. It feels like it changes the expectations, and in general it feels awkward as a CLI user.

Also, per that offline conversation, it sounds like the underlying hang getting in the way of a single Ctrl+C are being addressed, so I think I'm inclined to wait for those instead. We can revisit if needed.

Reply:
@liliankasem that's the default behavior. There are properties you can set in `Console` that would give you the ability to handle `Ctrl + C` multiple times, but it would be your responsibility to provide the termination handling.

Reply:
Resolving this thread as the scope of the PR has changed.